### PR TITLE
Implement xenserver vbd provider list attribute

### PR DESCRIPTION
**Pull request by acefei**
```
$make testacc
if [ -z "/home/feis/.local/go/bin" ]; then echo "GOBIN is not set" && exit 1; fi
go mod tidy
go install .
ls -l /home/feis/.local/go/bin/terraform-provider-xenserver
-rwxrwxr-x 1 feis feis 22730403 Jul  2 16:10 /home/feis/.local/go/bin/terraform-provider-xenserver
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkResource
--- PASS: TestAccNetworkResource (1.82s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (0.76s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.77s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (9.05s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (3.53s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (8.30s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (3.49s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (2.29s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (3.21s)
PASS
ok      terraform-provider-xenserver/xenserver  33.232s
```

```
Terraform will perform the following actions:

  # xenserver_vm.vm will be updated in-place
  ~ resource "xenserver_vm" "vm" {
      ~ hard_drive    = [
            "0119b54e-8305-4f01-9bbc-c6e19d7f1539",
          + "96c23cce-75d7-4593-9cbd-1fdd60e0bfb0",
        ]
        id            = "edcefc7a-ca80-1218-ab43-e045cd031cb2"
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Changes to Outputs:
  ~ vm_out = {
      ~ hard_drive    = [
            "0119b54e-8305-4f01-9bbc-c6e19d7f1539",
          + "96c23cce-75d7-4593-9cbd-1fdd60e0bfb0",
        ]
        id            = "edcefc7a-ca80-1218-ab43-e045cd031cb2"
        # (3 unchanged attributes hidden)
    }
xenserver_vm.vm: Modifying... [id=edcefc7a-ca80-1218-ab43-e045cd031cb2]
xenserver_vm.vm: Modifications complete after 0s [id=edcefc7a-ca80-1218-ab43-e045cd031cb2]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

Outputs:

vm_out = {
  "hard_drive" = tolist([
    "0119b54e-8305-4f01-9bbc-c6e19d7f1539",
    "96c23cce-75d7-4593-9cbd-1fdd60e0bfb0",
  ])
  "id" = "edcefc7a-ca80-1218-ab43-e045cd031cb2"
  "name_label" = "A test virtual-machine"
  "other_config" = tomap({})
  "template_name" = "Windows 11"
}
```

### CODE REVIEW

File: xenserver/vbd_utils.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	if len(userDevices) == 0 {
+		return "", errors.New("No available devices to attach to vm " + string(vmRef))
+	}
+
+	vbdRecord := xenapi.VBDRecord{
+		VM:         vmRef,
+		VDI:        vdiRef,
+		Type:       "Disk",
+		Mode:       "RW",
+		Bootable:   false,
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:27]: We may need to mention somewhere, the hard drives are not allow to boot.

File: xenserver/vbd_utils.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+func createVBD(vdiUUID string, vmRef xenapi.VMRef, session *xenapi.Session) (xenapi.VBDRef, error) {
+	vdiRef, err := xenapi.VDI.GetByUUID(session, vdiUUID)
+	if err != nil {
+		return "", errors.New("unable to get VDI ref, vdi UUID: " + vdiUUID)
+	}
+
+	userDevices, _ := xenapi.VM.GetAllowedVBDDevices(session, vmRef)
+	if len(userDevices) == 0 {
+		return "", errors.New("No available devices to attach to vm " + string(vmRef))
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:19]: "unbale to find available devices"

File: xenserver/vbd_utils.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	return vbdRef, nil
+}
+
+func createVBDs(ctx context.Context, vdiUUIDs types.List, vmRef xenapi.VMRef, session *xenapi.Session) ([]xenapi.VBDRef, error) {
+	elements := make([]string, 0, len(vdiUUIDs.Elements()))
+	diags := vdiUUIDs.ElementsAs(ctx, &elements, false)
+	if diags.HasError() {
+		return nil, errors.New("unable to get VDI UUIDs in plan data hard drive attributes")
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:57]: what about `unable to get VDI UUIDs from "hard_drive"` ?

File: xenserver/vbd_utils.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return "", errors.New("unable to create VBD, vdi UUID: " + vdiUUID)
+	}
+
+	// plug VBDs if VM is running
+	vmPowerState, err := xenapi.VM.GetPowerState(session, vmRef)
+	if err != nil {
+		return "", errors.New("unable to get VM power state, vm ref: " + string(vmRef))
+	}
+
+	if vmPowerState == xenapi.VMPowerStateRunning {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:43]: For vm in different Power State, do we have test about " vm add vbd on "Halted","Paused","Suspended" state", expect there is an XAPI error for not allowed operation, but we need to try.

File: xenserver/vbd_utils.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	diags := vdiUUIDs.ElementsAs(ctx, &elements, false)
+	if diags.HasError() {
+		return nil, errors.New("unable to get VDI UUIDs in plan data hard drive attributes")
+	}
+
+	var vbdRefs []xenapi.VBDRef
+	for _, vdiUUID := range elements {
+		vbdRef, err := createVBD(vdiUUID, vmRef, session)
+		if err != nil {
+			return nil, err
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:64]: return vbdRefs, err

File: xenserver/vbd_utils.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	"errors"
+	"xenapi"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+func createVBD(vdiUUID string, vmRef xenapi.VMRef, session *xenapi.Session) (xenapi.VBDRef, error) {
+	vdiRef, err := xenapi.VDI.GetByUUID(session, vdiUUID)
+	if err != nil {
+		return "", errors.New("unable to get VDI ref, vdi UUID: " + vdiUUID)
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:14]: we shouldn't return string type for xenapi.VBDRef.

return vdiRef, errors.New("unable to get VDI ref, vdi UUID: " + vdiUUID)

File: xenserver/vbd_utils.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		Type:       "Disk",
+		Mode:       "RW",
+		Bootable:   false,
+		Empty:      false,
+		Userdevice: userDevices[0],
+	}
+
+	vbdRef, err := xenapi.VBD.Create(session, vbdRecord)
+	if err != nil {
+		return "", errors.New("unable to create VBD, vdi UUID: " + vdiUUID)
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:34]: same like above

File: xenserver/vbd_utils.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	// plug VBDs if VM is running
+	vmPowerState, err := xenapi.VM.GetPowerState(session, vmRef)
+	if err != nil {
+		return "", errors.New("unable to get VM power state, vm ref: " + string(vmRef))
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:40]: return vbdRef, errors.New("unable to get VM power state")

File: xenserver/vbd_utils.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	if vmPowerState == xenapi.VMPowerStateRunning {
+		err = xenapi.VBD.Plug(session, vbdRef)
+		if err != nil {
+			return "", errors.New("unable to plug VBD, vdi UUID: " + vdiUUID)
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:46]: same

File: xenserver/vbd_utils.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+		planVDIsMap[vdiUUID] = true
+	}
+
+	stateVDIsMap := make(map[string]bool, len(stateVDIUUIDs))
+	for _, vdiUUID := range stateVDIUUIDs {
+		stateVDIsMap[vdiUUID] = true
+	}
+
+	// Create VBDs that are in plan but not in state
+	for vdiUUID := range planVDIsMap {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:85]: what about use "slice" packages for this?

File: xenserver/vbd_utils.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+		if _, ok := stateVDIsMap[vdiUUID]; !ok {
+			_, err = createVBD(vdiUUID, vmRef, session)
+			if err != nil {
+				return err
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:89]: if we return err here, I don't think we need "var err error"

File: xenserver/vbd_utils.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+func removeVBDbyVDIUUID(session *xenapi.Session, vdiUUID string) error {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:107]: removeVBDbyVDIUUID -> removeVBDByVDIUUID

File: xenserver/vbd_utils.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Destroy VBDs that are not in plan
+	for vdiUUID := range stateVDIsMap {
+		if _, ok := planVDIsMap[vdiUUID]; !ok {
+			err = removeVBDbyVDIUUID(session, vdiUUID)
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:97]: little confuse here.  do we need remove all vbd on this vdi or just for this VM?

File: xenserver/vm_resouce.go [hunk 14]
```diff
@@ -133,6 +80,15 @@ func (r *vmResource) Create(ctx context.Context, req resource.CreateRequest, res
 		return
 	}
 
+	err = xenapi.VM.SetIsATemplate(r.session, vmRef, false)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to set VM from a template",
```
Comment (xueqingz) [inline xenserver/vm_resouce.go:86]: `Unable to set VM as "not a template"`

File: xenserver/vm_resouce.go [hunk 15]
```diff
@@ -242,31 +192,34 @@ func (r *vmResource) Update(ctx context.Context, req resource.UpdateRequest, res
...
 		resp.Diagnostics.AddError(
 			"Unable to get VM ref",
 			err.Error(),
 		)
 		return
 	}
+
 	// Update existing vm resource with new plan
-	err = xenapi.VM.SetNameLabel(r.session, vmRef, data.NameLabel.ValueString())
+	err = xenapi.VM.SetNameLabel(r.session, vmRef, plan.NameLabel.ValueString())
```
Comment (xueqingz) [inline xenserver/vm_resouce.go:205]: we'd better move all those "update" logic to vm_utils in future to make vm_resource.go very clear and clean.

File: xenserver/vbd_utils.go (hunk 4 above)
Comment (acefei) [reply xenserver/vbd_utils.go:43]: Just follow the logic you sent about XCM logic.

File: xenserver/vm_resouce.go (hunk 15 above)
Comment (acefei) [reply xenserver/vm_resouce.go:205]: The update logic might not be necessary moved to vm_utils, considering the function is named "Update".

Comment (xueqingz) [reply xenserver/vm_resouce.go:205]: I mean, In resource update function, we do 1. check if the parameters can be updated 2. update the resources 3. get new state and store in state.  I want those steps looks clear with 3 methods if possible.

File: xenserver/vbd_utils.go (hunk 11 above)
Comment (acefei) [reply xenserver/vbd_utils.go:89]: there are many errors when I removed "var err error"

File: xenserver/vbd_utils.go (hunk 13 above)
Comment (acefei) [reply xenserver/vbd_utils.go]: add a check if VMref matched.

File: xenserver/vm_utils.go [hunk 16]
```diff
@@ -323,29 +367,67 @@ func getVMOtherConfig(ctx context.Context, data vmResourceModel) (map[string]str
...
 	delete(vmRecord.OtherConfig, "template_name")
+	var diags diag.Diagnostics
 	data.OtherConfig, diags = types.MapValueFrom(ctx, types.StringType, vmRecord.OtherConfig)
 	if diags.HasError() {
 		return errors.New("unable to read VM other config")
 	}
-	err := updateVMResourceModelComputed(ctx, vmRecord, data)
+
+	err := updateVMResourceModelComputed(ctx, session, vmRecord, data)
 	if err != nil {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:393]: "return err" or  "return updateVMResourceModelComputed(ctx, session, vmRecord, data)"

File: xenserver/vbd_utils.go [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+			MarkdownDescription: "Set VBD as bootable, Default: false",
+			Optional:            true,
+			Computed:            true,
+			Default:             booldefault.StaticBool(false),
+		},
+		"mode": schema.StringAttribute{
+			MarkdownDescription: "The mode the VBD should be mounted with, Default: RW",
+			Optional:            true,
+			Computed:            true,
+			Default:             stringdefault.StaticString("RW"),
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:46]: Do we need a filed "vbd_uuid" here?

File: xenserver/vbd_utils.go [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+		},
+	}
+}
+
+func createVBD(vbd vbdResourceModel, vmRef xenapi.VMRef, session *xenapi.Session) (xenapi.VBDRef, error) {
+	var vbdRef xenapi.VBDRef
+	vdiRef, err := xenapi.VDI.GetByUUID(session, vbd.VDI.ValueString())
+	if err != nil {
+		return vbdRef, errors.New("unable to get VDI ref, vdi UUID: " + vbd.VDI.String())
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:55]: expect to return Xapi error than error string.

File: xenserver/vbd_utils.go (hunk 17 above)
Comment (acefei) [reply xenserver/vbd_utils.go:46]: what is it used for ?

File: xenserver/vm_resource_test.go [hunk 19]
```diff
@@ -28,11 +45,11 @@ func TestAccVMResource(t *testing.T) {
 				Config: providerConfig + testAccVMResourceConfig("test vm 1"),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					resource.TestCheckResourceAttr("xenserver_vm.test_vm", "name_label", "test vm 1"),
-					resource.TestCheckResourceAttr("xenserver_vm.test_vm", "template_name", "CentOS 7"),
+					resource.TestCheckResourceAttr("xenserver_vm.test_vm", "template_name", "Windows 11"),
+					resource.TestCheckResourceAttr("xenserver_vm.test_vm", "hard_drive.#", "1"),
```
Comment (xueqingz) [inline xenserver/vm_resource_test.go:49]: we can add more test cases here, eg.
"resource.TestCheckResourceAttr("xenserver_vm.test_vm", "hard_drive.0.%", "4")," 
"resource.TestCheckResourceAttr("xenserver_vm.test_vm", "hard_drive.0.mode", "RW")," 
"resource.TestCheckResourceAttr("xenserver_vm.test_vm", "hard_drive.0.bootable", "true")," 
resource.TestCheckResourceAttrSet("xenserver_vm.test_vm", "hard_drive.0.vbd_ref"),"

**Review by xueqingz — APPROVED** [review]

**Review by xueqingz — CHANGES_REQUESTED** [review]
Do we missing the change for  terraform-provider-xenserver/examples/resources/xenserver_vm/resource.tf ?

**Review by xueqingz — APPROVED** [review]

**Review by xihuan-citrix — APPROVED** [review]

**Pull request merged**